File: sdk-typescript/src/WarmPool.ts
/*
 * Copyright Daytona Platforms Inc.
 * SPDX-License-Identifier: Apache-2.0
 */

import { WarmPoolsApi } from '@daytona/api-client'
import type { CreateWarmPool, UpdateWarmPool, WarmPool as WarmPoolDto } from '@daytona/api-client'
import { WithInstrumentation } from './utils/otel.decorator'

/**
 * Represents a Daytona Warm Pool which keeps ready-to-use Sandboxes for a snapshot.
 *
 * `currentSize` versus `pool` is the pool's status: `currentSize` is the number of ready
 * warm sandboxes, `pool` is the desired number. `errorReason` is set when the pool cannot
 * be filled.
 *
 * @property {string} id - Unique identifier for the Warm Pool
 * @property {string} organizationId - Organization ID that owns the Warm Pool
 * @property {string} snapshot - Snapshot the pool keeps warm sandboxes for
 * @property {string} target - Target region of the pool
 * @property {number} pool - Desired number of warm sandboxes
 * @property {number} currentSize - Current number of ready warm sandboxes in the pool
 * @property {number} cpu - CPU cores per sandbox
 * @property {number} mem - Memory per sandbox in GiB
 * @property {number} disk - Disk per sandbox in GiB
 * @property {string} osUser - OS user of the warm sandboxes
 * @property {Record<string, string>} env - Environment variables of the warm sandboxes
 * @property {string | null} [errorReason] - Reason the pool cannot be filled, if any
 * @property {string} createdAt - Date and time when the Warm Pool was created
 * @property {string} updatedAt - Date and time when the Warm Pool was last updated
 */
export type WarmPool = WarmPoolDto & { __brand: 'WarmPool' }

/**
 * Service for managing Daytona Warm Pools.
 *
 * This service provides methods to list, create, update, and delete Warm Pools.
 *
 * @class
 */
export class WarmPoolService {
  constructor(private warmPoolsApi: WarmPoolsApi) {}

  /**
   * Lists all Warm Pools in the organization.
   *
   * @returns {Promise<WarmPool[]>} List of all Warm Pools
   *
   * @example
   * const daytona = new Daytona();
   * const pools = await daytona.warmPool.list();
   * pools.forEach(pool => console.log(`${pool.snapshot}: ${pool.currentSize}/${pool.pool} ready`));
   */
  @WithInstrumentation()
  async list(): Promise<WarmPool[]> {
    const response = await this.warmPoolsApi.listWarmPools()
    return response.data as WarmPool[]
  }

  /**
   * Creates a new Warm Pool.
   *
   * @param {CreateWarmPool} params - Parameters for the new Warm Pool
   * @returns {Promise<WarmPool>} The newly created Warm Pool
   * @throws {DaytonaConflictError} If a Warm Pool for the same snapshot and region already exists
   *
   * @example
   * const daytona = new Daytona();
   * const pool = await daytona.warmPool.create({ snapshot: 'my-snapshot', pool: 5 });
   * console.log(`Created warm pool ${pool.id} in ${pool.target}`);
   */
  @WithInstrumentation()
  async create(params: CreateWarmPool): Promise<WarmPool> {
    const response = await this.warmPoolsApi.createWarmPool(params)
    return response.data as WarmPool
  }

  /**
   * Updates the desired size of a Warm Pool.
   *
   * @param {string} warmPoolId - ID of the Warm Pool to update
   * @param {UpdateWarmPool} params - Fields to update (`pool: 0` drains the pool)
   * @returns {Promise<WarmPool>} The updated Warm Pool
   * @throws {DaytonaNotFoundError} If the Warm Pool does not exist
   *
   * @example
   * const daytona = new Daytona();
   * const pool = await daytona.warmPool.update("warm-pool-id", { pool: 10 });
   */
  @WithInstrumentation()
  async update(warmPoolId: string, params: UpdateWarmPool): Promise<WarmPool> {
    const response = await this.warmPoolsApi.updateWarmPool(warmPoolId, params)
    return response.data as WarmPool
  }

  /**
   * Deletes a Warm Pool.
   *
   * @param {string} warmPoolId - ID of the Warm Pool to delete
   * @returns {Promise<void>}
   * @throws {DaytonaNotFoundError} If the Warm Pool does not exist
   *
   * @example
   * const daytona = new Daytona();
   * await daytona.warmPool.delete("warm-pool-id");
   */
  @WithInstrumentation()
  async delete(warmPoolId: string): Promise<void> {
    await this.warmPoolsApi.deleteWarmPool(warmPoolId)
  }
}

File: sdk-python/src/daytona/_async/warm_pool.py
# Copyright Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from daytona_api_client_async import CreateWarmPool, UpdateWarmPool, WarmPoolsApi

from .._utils.otel_decorator import with_instrumentation
from ..common.warm_pool import WarmPool


class AsyncWarmPoolService:
    """Service for managing Daytona Warm Pools. Can be used to list, create, update and delete Warm Pools."""

    def __init__(self, warm_pools_api: WarmPoolsApi):
        self.__warm_pools_api = warm_pools_api

    @with_instrumentation()
    async def list(self) -> list[WarmPool]:
        """List all Warm Pools in the organization.

        Returns:
            list[WarmPool]: List of all Warm Pools.

        Example:
            ```python
            async with AsyncDaytona() as daytona:
                pools = await daytona.warm_pool.list()
                for pool in pools:
                    print(f"{pool.snapshot}: {pool.current_size}/{pool.pool} ready")
            ```
        """
        return [WarmPool.from_dto(pool) for pool in await self.__warm_pools_api.list_warm_pools()]

    @with_instrumentation()
    async def create(self, snapshot: str, pool: int, target: str | None = None) -> WarmPool:
        """Create a new Warm Pool.

        Args:
            snapshot (str): The snapshot (ID or name) to keep warm sandboxes for.
            pool (int): Number of warm sandboxes to keep ready.
            target (str | None): Target region for the Warm Pool. Defaults to the
                organization default region.

        Returns:
            WarmPool: The newly created Warm Pool.

        Raises:
            ApiException: If a Warm Pool for the same snapshot and region already exists (409).

        Example:
            ```python
            async with AsyncDaytona() as daytona:
                pool = await daytona.warm_pool.create("my-snapshot", pool=5)
                print(f"Created warm pool {pool.id} in {pool.target}")
            ```
        """
        return WarmPool.from_dto(
            await self.__warm_pools_api.create_warm_pool(CreateWarmPool(snapshot=snapshot, pool=pool, target=target))
        )

    @with_instrumentation()
    async def update(self, warm_pool_id: str, pool: int) -> WarmPool:
        """Update the desired size of a Warm Pool.

        Args:
            warm_pool_id (str): ID of the Warm Pool to update.
            pool (int): New desired number of warm sandboxes (0 drains the pool).

        Returns:
            WarmPool: The updated Warm Pool.

        Raises:
            NotFoundException: If the Warm Pool does not exist.

        Example:
            ```python
            async with AsyncDaytona() as daytona:
                pool = await daytona.warm_pool.update("warm-pool-id", pool=10)
            ```
        """
        return WarmPool.from_dto(await self.__warm_pools_api.update_warm_pool(warm_pool_id, UpdateWarmPool(pool=pool)))

    @with_instrumentation()
    async def delete(self, warm_pool_id: str) -> None:
        """Delete a Warm Pool.

        Args:
            warm_pool_id (str): ID of the Warm Pool to delete.

        Raises:
            NotFoundException: If the Warm Pool does not exist.

        Example:
            ```python
            async with AsyncDaytona() as daytona:
                await daytona.warm_pool.delete("warm-pool-id")
                print("Warm pool deleted")
            ```
        """
        await self.__warm_pools_api.delete_warm_pool(warm_pool_id)
